### ISSUE TITLE
ignore @Generated, and a config option to filter by annotation

### ISSUE THREAD

**Issue by xenoterracide** [spotbugs/spotbugs#1385]
I think that spotbugs should auto ignore (or be easily configurable to) classes/methods/fields decorated with `@javax.annotation.processing.Generated`, also some frameworks don't use this, so honestly I think it should ignore those things if the name of the annotation is `Generated`. Alternatively, maybe just add an option to the exclude/filter config, e.g.

```xml
    <Match>
        <Bug code="L,D,BC"/>
        <Annotation name="org.immutables.value.Generated"/>
    </Match>
```

**Comment by marquiswang** [comment]
Would love to have this as well. Using Dagger and Immutables, which lead to many false positives (or at least, not fixable by me positives).

**Comment by Lucas3oo** [comment]
+1 same with Micronaut generated code. Those have @io.micronaut.core.annotation.Generated

**Comment by RSLak1** [comment]
Just wanted to start working on this and found that it has already been done. This issue should be covered by https://github.com/spotbugs/spotbugs/pull/2395 and should therefore be included in the next release.

As far as I can see, the corresponding documentation in docs/filter.rst is missing though. I could add some, are there other places where they would be needed? @hazendaz

**Comment by hazendaz** [comment]
@RenovatingDev Not sure where all documentation is missing, please do send PR for anything you find.

**Comment by sergeykad** [comment]
It works with version 4.8.0.

```xml
<?xml version="1.0" encoding="UTF-8"?>
<FindBugsFilter
        xmlns="https://github.com/spotbugs/filter/3.0.0"
        xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
        xsi:schemaLocation="https://github.com/spotbugs/filter/3.0.0 https://raw.githubusercontent.com/spotbugs/spotbugs/4.8.0/spotbugs/etc/findbugsfilter.xsd">
    <Match>
        <Annotation name="javax.annotation.processing.Generated"/>
    </Match>
</FindBugsFilter>
```

**Comment by gim-** [comment]
@sergeykad I've been looking for something like this, the feature is not documented anywhere, and the link really helped. Thank you!

The reason it's not working for you is because you're not utilizing it correctly. Here's a working example:
```xml
<Match>
    <Bug code="RCN" classAnnotationNames="lombok.EqualsAndHashCode"/>
</Match>
```
This disables RCN checks for classes annotated with `lombok.EqualsAndHashCode`.

**Comment by xenoterracide** [comment]
seems like it's working for me, 

```
<?xml version="1.0" encoding="UTF-8" ?>
<FindBugsFilter
  xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
  xmlns="https://github.com/spotbugs/filter/3.0.0"
  xsi:schemaLocation="https://github.com/spotbugs/filter/3.0.0  https://raw.githubusercontent.com/spotbugs/spotbugs/4.8.0/spotbugs/etc/findbugsfilter.xsd"
>
  <Match>
    <!-- Returning mutable references, I slightly agree but this is also ludicrous level -->
    <Bug pattern="EI_EXPOSE_REP" />
  </Match>
  <Match>
    <Bug pattern="UWF_FIELD_NOT_INITIALIZED_IN_CONSTRUCTOR" />
    <Class name="~.*Test.*" />
  </Match>
  <Match>
    <!-- safe to ignore in tests due to gitattributes -->
    <Bug pattern="VA_FORMAT_STRING_USES_NEWLINE" />
    <Class name="~.*Test$?.*" />
  </Match>
  <Match>
    <Annotation name="jakarta.persistence.metamodel.StaticMetamodel" />
  </Match>
</FindBugsFilter>
```

**Comment by tlefevre** [comment]
Doesn't work for me either. It seems it only works if the class itself has the annotation and won't ignore generated methods with a Generated annotation for example.

**Comment by Vampire** [comment]
Yes, that's correct.
That's what the above linked release notes tell.
If you want it to match on methods, I guess you should post a separate feature request.

**Comment by rcolombo** [comment]
This also doesn't work for me. I'm trying to ignore `javax.annotation.processing.Generated` at the class-level. I have got it to work for other annotations though.

**Comment by Vampire** [comment]
That annotation has `SOURCE` retention, but Spotbugs is reading annotations from the class files.
You need an annotation with at least `CLASS` retention for this to work.

**Comment by sergeykad** [comment]
You are correct, but this is the most useful use case for filtering by annotation.

Our project for example has multiple locations with generated code. Without the option to filter by annotation the only available option is to find all directories with the generated code and list them in exclusions one by one.

**Comment by Vampire** [comment]
You can filter by annotation, just make sure it has an appropriate one. Many generation tools you can configure which annotation to add. For other cases where the annotation is hard-code I usually add a `doLast` action to the generation task that adds the proper annotation that can be used for that filter to the generated files.

**Comment by sergeykad** [comment]
You are correct, but this is the most useful use case for filtering by annotation.

Our project for example has multiple locations with generated code. Without the option to filter by annotation the only available option is to find all directories with the generated code and list them in exclusions one by one.

I can be mistaken, but I don't remember Spotbugs documentation mentioning annotation retention type.

**Comment by sergeykad** [comment]
@Vampire What annotation would you recommend for the generated code?

**Comment by Vampire** [comment]
> I can be mistaken, but I don't remember Spotbugs documentation mentioning annotation retention type.

You might be right, but it just seems logical, as you give class files to Spotbugs for analysis.
And I had a quick look at the code, where the annotation information is coming through BCEL which is a a library for Bytecode.
You can optionally give it source files, but those are only used for showing code in the report iirc.
You could of course open an improvement issue to also support source-retention annotations if source is given. :-)

> What annotation would you recommend for the generated code?

You really can use any that does not have source retention.
There are some `Generated`-like annotations out there that have other than source retention, but usually they are part of some bigger library and I don't want to add all that to my classpath unless I have it anyway already.
So most often I just create an own annotation with simple name `Generated` and use that.

**Comment by xenoterracide** [comment]
huh, remind me to file a bug on jakarta's annotation :/

Anyways, in addition creating your own, sometimes the generated code includes another annotation. Immutables for example has it's own Generated annotation which can be applied in addition to jakarta. For hibernates metamodel though I put this rule in. While it still has jakarta's generated, that annotation is runtime.

```
  <Match>
    <Annotation name="jakarta.persistence.metamodel.StaticMetamodel" />
  </Match>
```

**Comment by xenoterracide** [comment]
I think this is fixed...

**Comment by xenoterracide** [comment]
There's already an open issue about this in Jakarta, I'd say feel free to comment and vote

**Comment by alabotski** [comment]
Seems this match doesnt'work

```
<Match>
   <Annotation name="jakarta.annotation.Generated"/>
</Match>
```

**Comment by Vampire** [comment]
Of course not, see https://github.com/spotbugs/spotbugs/issues/1385#issuecomment-2264123711
